src/purview.ts

```diff
@@ -1189,57 +914,332 @@ const globalStateTrees: Record<string, StateTree | undefined> = {}
+      if (!component) {
+        throw new Error("Expected non-null component")
+      }
+
+      let onUnseenError: ErrorHandler | null = null
+      if (onError) {
+        onUnseenError = error => {
+          if (typeof error !== "object" || error === null) {
+            onError(error)
+            return
+          }
+
+          if (!seenErrors.has(error)) {
+            seenErrors.add(error)
+            onError(error)
+          }
+        }
+      }
+
+      let root: ConnectedRoot | DisconnectedRoot
+      if (purviewState) {
+        // This is the request from the websocket connection.
+        component._id = idStateTree!.id
+        root = {
+          connected: true,
+          component,
+          wsState: purviewState.wsState,
+          eventNames: new Set(),
+          aliases: {},
+          allComponentsMap: { [component._id]: component },
+          onError: onUnseenError,
+        }
+        purviewState.roots = purviewState.roots || []
+        purviewState.roots.push(root)
+      } else {
+        // This is the initial render.
+        req.purviewCSSState = req.purviewCSSState ?? {
+          id: nanoid(),
+          atomicCSS: {},
+          cssRules: [],
+          nextRuleIndex: 0,
+        }
+        if (req.purviewCSSRendered) {
+          throw new Error(RENDER_CSS_ORDERING_ERROR)
+        }
+        root = {
+          connected: false,
+          cssState: req.purviewCSSState,
+          onError: onUnseenError,
+        }
+      }
+
+      const pNode = await renderComponent(component, component._id, root)
+      if (!pNode) {
+        throw new Error("Expected non-null node")
+      }
 
-  async deleteStateTree(id: string): Promise<void> {
-    delete globalStateTrees[id]
-  },
+      if (purviewState) {
+        return ""
+      } else {
+        await reloadOptions.saveStateTree(
+          component._id,
+          makeStateTree(component, false),
+        )
+        return toHTML(pNode)
+      }
+    })
+  }
 
-  async saveCSSState(id: string, cssState: CSSState): Promise<void> {
-    globalCSSState[id] = cssState
-    if (process.env.NODE_ENV !== "test") {
-      setTimeout(() => this.deleteCSSState(id), DELETE_INTERVAL)
+  export async function renderCSS(req: http.IncomingMessage): Promise<string> {
+    const cssState = req.purviewCSSState
+    if (!cssState) {
+      req.purviewCSSRendered = true
+      return ""
     }
-  },
 
-  async getCSSState(id: string): Promise<CSSState | null> {
-    return globalCSSState[id] ?? null
-  },
+    const { id, cssRules } = cssState
+    const textPNode = createTextPNode(cssRules.join("\n"))
+    const pNode = createPNode(
+      "style",
+      { id: STYLE_TAG_ID, "data-css-state-id": id },
+      [textPNode],
+    )
+
+    cssState.nextRuleIndex = cssRules.length
+    await reloadOptions.saveCSSState(id, cssState)
+    req.purviewCSSRendered = true
+    return toHTML(pNode)
+  }
+  export import JSX = JSXInternal
```

Comment:
This is the all-important line. With this syntax the JSX namespace as defined by `JSXInternal` comes to live on the `Purview` namespace, allowing TypeScript to find `Purview.JSX` when it applies the JSX factory `Purview.createElem` to Purview code.